packages/kotti-ui/source/props.test.ts

```diff
@@ -0,0 +1,451 @@
+import { PropOptions } from '@vue/composition-api'
+import { z } from 'zod'
+
+import { makeProps } from './props'
+import { silenceConsole } from './test-utils/silence-console'
+
+const expectDefault = (
+	prop: PropOptions<unknown, boolean>,
+	defaultValue: unknown,
+) => {
+	// @ts-expect-error
+	expect(prop.default()).toEqual(defaultValue)
+	expect(prop.required).toBe(undefined)
+}
+
+const expectRequired = (prop: PropOptions<unknown, boolean>) => {
+	expect(prop.default).toEqual(undefined)
+	expect(prop.required).toBe(true)
+}
+
+const expectValidator = (
+	prop: PropOptions<unknown, boolean>,
+	successCases: unknown[],
+	failCases: unknown[],
+) => {
+	const { validator } = prop
+
+	if (validator === undefined) {
+		expect(validator).not.toBe(undefined)
+		throw new Error('unexpected undefined validator')
+	}
+
+	successCases.forEach((c) => expect(validator(c)).toBe(true))
+	failCases.forEach((c) => expect(validator(c)).toBe(false))
+}
+
+describe('array', () => {
+	beforeAll(silenceConsole)
+
+	it('z.array(z.number())', () => {
+		const schema = z.object({
+			prop: z.array(z.number()),
+		})
+		const { prop } = makeProps(schema)
+
+		expectRequired(prop)
+		expect(prop.type).toBe(Array)
+		expectValidator(prop, [[], [1], [-1]], [1, null, 'test', undefined])
+	})
+
+	it('z.array(z.number()).nullable()', () => {
```

Comment:
likewise you can have the generic suite for `nullable` where you clarify that nullable unions the type to the general type, and not to each element

so `nullable` array, would add Array<X> | null BUT NOT Array<X | null>

Reply:
up to you

Reply:
Somewhat addressed via `TYPE_SUCCESS`/`TYPE_FAILURE`